src/main/java/com/conveyal/datatools/manager/controllers/api/LabelController.java

```diff
@@ -0,0 +1,224 @@
+                        .collect(Collectors.toList());
+                // If there are changes, update the feed source
+                if (!newLabels.equals(feedSource.labels)) {
+                    FeedSource fs = Persistence.feedSources.getById(feedSource.id);
+                    fs.labels = newLabels;
+                    Persistence.feedSources.replace(fs.id, fs);
+                }
+            });
+
+            label.delete();
+            return label;
+        } catch (Exception e) {
+            logMessageAndHalt(req, 500, "Unknown error occurred while deleting label.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Helper function returns label if user has permission for specified action.
+     * @param req spark Request object from API request
+     * @param action action type (either "view" or Permission.MANAGE)
+     * @return label object for ID
+     */
+    private static Label requestLabelById(Request req, Actions action) {
+        String id = req.params("id");
+        if (id == null) {
+            logMessageAndHalt(req, 400, "Please specify id param");
+        }
+        return checkLabelPermissions(req, Persistence.labels.getById(id), action);
+    }
+
+    /** Helper method returns label only if user is authorized to do given action with it
+     *
+     * @param req       Spark request used for determining user permissions
+     * @param label     Label to check
+     * @param action    Action to be taken on label, changes who can do what
+     * @return          Label which is ok to return to user
+     */
+    public static Label checkLabelPermissions(Request req, Label label, Actions action) {
+        Auth0UserProfile userProfile = req.attribute("user");
+        // check for null label
+        if (label == null) {
+            logMessageAndHalt(req, 400, "Label ID does not exist");
+            return null;
+        }
+
+        boolean authorized;
+        boolean isProjectAdmin = userProfile.canAdministerProject(label.projectId, label.organizationId());
+        switch (action) {
+            case CREATE:
+            case MANAGE:
+            case EDIT:
+                // Only project adminds can edit labels
+                authorized = isProjectAdmin;
+                break;
+            case VIEW:
+                authorized = true;
+                if (label.adminOnly) {
+                    authorized = isProjectAdmin;
+                }
+                break;
+            default:
+                authorized = false;
+                break;
+        }
+
+        if (!authorized) {
+            // Special message if viewing an adminOnly label
+            if (label.adminOnly) {
+                logMessageAndHalt(req, 403, "User is not admin so cannot view admin-only label");
+            }
+            // Throw halt if user not authorized.
+            logMessageAndHalt(req, 403, "User is not admin so cannot update or create labels");
+        }
```

Comment:
These `logMessageAndHalt`s could be handled within the above switch statement. Also, the `if(label.adminOnly)` could also apply to the `CREATE`, `MANAGE` or `DELETE` cases, so the message may not be correct anyways.

```suggestion
        boolean isProjectAdmin = userProfile.canAdministerProject(label.projectId, label.organizationId());
        switch (action) {
            case CREATE:
            case MANAGE:
            case EDIT:
                // Only project admins can edit labels
                logMessageAndHalt(req, 403, "User is not admin so cannot update or create labels");
                break;
            case VIEW:
                if (label.adminOnly && !isProjectAdmin) {
                    logMessageAndHalt(req, 403, "User is not admin so cannot view admin-only label");
                }
                break;
        }
```

Reply:
This looks much cleaner thank you! Committed in 
85261da